### PR TITLE
Added whoosh search backend and realtime indexing

### LINKED ISSUE #325: Enable more robust search backend with realtime indexing
Searching the forums with the default Haystack search backend in django-machina does not reliably return search results. According to the [Haystack docs](https://django-haystack.readthedocs.io/en/v2.7.0/tutorial.html#simple) this is not unexpected behavior for the very barebones 'simple' backend implementation, which is not recommended for use in production. To that end, the purpose of this issue is to enable the more robust 'whoosh' backend that is recommended in the [django-machina docs](https://django-machina.readthedocs.io/en/stable/getting_started.html#django-haystack-settings) as an alternative to the default.  Additionally, to alleviate the need to run `update_index` every time the forum content changes to make it searchable, Haystack's [realtime signal processor ](https://django-haystack.readthedocs.io/en/latest/signal_processors.html#realtime-realtimesignalprocessor) will also need to be activated.

### PR DESCRIPTION
**Pull request by joshuamonroe**
Per issue #325, I've enabled the much more robust whoosh search backend in django-machina, with realtime indexing enabled so that we don't have run `python3 mangage.py update_index` every time the forum content changes. Note, that you will still need to run this command once to index any pre-existing forum content.

This backend generates index files which are stored in `src/chigame/forums/search_index`. I've added this directory to `.gitignore` and while I was in there I also added the `.vscode` settings directory.

### CODE REVIEW

**Review by Benjoman12 — APPROVED** [review]
This looks good to me as it both cleanly adds the necessary items and adds clarity to the settings file regarding the django-machina related items.

File: src/config/settings/base.py [hunk 1]
```diff
@@ -322,17 +322,19 @@
...
 # https://django-machina.readthedocs.io/en/latest/getting_started.html#django-haystack-settings
 HAYSTACK_CONNECTIONS = {
     "default": {
-        "ENGINE": "haystack.backends.simple_backend.SimpleEngine",
+        "ENGINE": "haystack.backends.whoosh_backend.WhooshEngine",
+        "PATH": str(BASE_DIR / "chigame/forums/search_index"),
     },
 }
 
+# https://django-haystack.readthedocs.io/en/latest/signal_processors.html
```
Comment (majorsylvie) [inline src/config/settings/base.py:353]: I would add a comment briefly mentioning what a "signal processor" is and why you chose this one.

File: .gitignore [hunk 2]
```diff
@@ -159,9 +159,15 @@ cython_debug/
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 .idea/
 
+# Visual Studio Code
+.vscode/
+
 # Misago Sandbox Demo Repo
 src/sandbox/misago-demo/Misago
 
+# Django-machina whoosh search index
```
Comment (majorsylvie) [inline .gitignore]: I would add a comment basically saying what you did in the summary that files are generated by the indexer and put here based on the settings. 

Please link to relevant documentation, expand on why this is the specific directory, and why it should be in the gitignore

File: .gitignore [hunk 3]
```diff
@@ -159,9 +159,15 @@ cython_debug/
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 .idea/
 
+# Visual Studio Code
```
Comment (majorsylvie) [inline .gitignore:162]: similar to below, add a comment explaining what files vscode generates that we don't want in git.

**Review by majorsylvie — CHANGES_REQUESTED** [review]
Overall, excited to see changes like this!

Please add more comments explaining your code, specific guidance in-review comments.

FYI, it looks like the actual search only uses full-string matches instead of substring:

image

image

image
image
image

You could consider adding substring search support, though I do not know what that would take in terms of indexing.

Please add further comments and then this PR should be good to go!

**Comment by joshuamonroe** [comment]
@majorsylvie I've added the additional documentation that you requested.

Good catch on the substring matching. I looked into it quickly and, although I couldn't find anything definitive on how to implement it, it does look like it would likely require changing how content is indexed in django-machina. I can create an issue in the backlog for future devs to address this.

File: .gitignore (hunk 3 above)
Comment (majorsylvie) [reply .gitignore:162]: phenomenal comment!

File: .gitignore [hunk 4]
```diff
@@ -159,6 +159,23 @@ cython_debug/
...
+#  https://code.visualstudio.com/docs/getstarted/settings#_workspace-settings
+.vscode/
+
+# Django-machina Forum Search Index
+#  This directory contains index files generated by django-machina's Whoosh
+#  search backend. These files are uniquely generated based on the the content
+#  of a given forum instance and should not be commited to version control.
+#
+#  The location of this directory is set here:
+#   settings/base.py > DJANGO-MACHINA SETTINGS > HAYSTACK_CONNECTIONS > "PATH"
```
Comment (majorsylvie) [inline .gitignore:175]: Incredibly descriptive and relevant comment. Wonderful job!

File: src/config/settings/base.py (hunk 1 above)
Comment (majorsylvie) [reply src/config/settings/base.py:353]: Please make an issue for this queued search and put it in the backlog!

**Review by majorsylvie — APPROVED** [review]
**Please make an issue for implementing a queued search and put it in the backlog!**

Absolutely incredibly job with the code comments. 
Very descriptive, focused, clearly written, with proper references to documentation and relevant concepts.

Phenomenal work! 
This is good to merge and wonderful work.

**Comment by joshuamonroe** [comment]
I've added the following issue to implement queued search at a later date.
- #416 

Thanks to @Benjoman12 for reviewing the PR.

**Comment by majorsylvie** [comment]
Issue Score: Excellent

Comments:
The title is too generic and could relate to other teams.

**Pull request merged**